File: src/main/java/org/example/kalkulationsprogramm/config/AsyncConfig.java
package org.example.kalkulationsprogramm.config;

import java.util.concurrent.Executor;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.context.annotation.Primary;
import org.springframework.core.task.TaskExecutor;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.scheduling.TaskScheduler;
import org.springframework.scheduling.concurrent.ThreadPoolTaskExecutor;
import org.springframework.scheduling.concurrent.ThreadPoolTaskScheduler;

@Configuration
public class AsyncConfig {

    @Bean(name = "emailTaskExecutor")
    public TaskExecutor emailTaskExecutor() {
        ThreadPoolTaskExecutor executor = new ThreadPoolTaskExecutor();
        executor.setCorePoolSize(4);
        executor.setMaxPoolSize(8);
        executor.setQueueCapacity(200);
        executor.setThreadNamePrefix("email-sync-");
        executor.initialize();
        return executor;
    }

    /**
     * Default-Executor für alle @Async-Methoden ohne expliziten Namen.
     * Verhindert, dass Spring auf SimpleAsyncTaskExecutor zurückfällt
     * (der pro Aufruf einen neuen Thread spawnt) oder den taskScheduler belastet.
     */
    @Bean(name = "defaultTaskExecutor")
    @Primary
    public ThreadPoolTaskExecutor taskExecutor() {
        ThreadPoolTaskExecutor executor = new ThreadPoolTaskExecutor();
        executor.setCorePoolSize(4);
        executor.setMaxPoolSize(16);
        executor.setQueueCapacity(500);
        executor.setThreadNamePrefix("async-");
        executor.setWaitForTasksToCompleteOnShutdown(true);
        executor.setAwaitTerminationSeconds(30);
        executor.initialize();
        return executor;
    }

    @Bean(name = "applicationTaskExecutor")
    public Executor applicationTaskExecutor(@Qualifier("defaultTaskExecutor") ThreadPoolTaskExecutor executor) {
        return executor.getThreadPoolExecutor();
    }

    /**
     * TaskScheduler für @Scheduled-Methoden.
     * Default-Pool wäre 1 Thread → ein hängender @Scheduled-Job blockiert
     * alle anderen. Mit 4 Threads bleibt der Email-Import unabhängig von
     * anderen periodischen Jobs (Cleanup, WebPush, SpamBayes etc.).
     */
    @Bean(name = "taskScheduler")
    public TaskScheduler taskScheduler() {
        ThreadPoolTaskScheduler scheduler = new ThreadPoolTaskScheduler();
        scheduler.setPoolSize(4);
        scheduler.setThreadNamePrefix("scheduled-");
        scheduler.setWaitForTasksToCompleteOnShutdown(true);
        scheduler.setAwaitTerminationSeconds(30);
        scheduler.initialize();
        return scheduler;
    }
}

File: src/main/kotlin/org/example/kalkulationsprogramm/config/AsyncConfig.kt
package org.example.kalkulationsprogramm.config

import org.springframework.beans.factory.annotation.Qualifier
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.context.annotation.Primary
import org.springframework.core.task.TaskExecutor
import org.springframework.scheduling.TaskScheduler
import org.springframework.scheduling.concurrent.ThreadPoolTaskExecutor
import org.springframework.scheduling.concurrent.ThreadPoolTaskScheduler
import java.util.concurrent.Executor

@Configuration
class AsyncConfig {

    @Bean(name = ["emailTaskExecutor"])
    fun emailTaskExecutor(): TaskExecutor {
        val executor = ThreadPoolTaskExecutor()
        executor.corePoolSize = 4
        executor.maxPoolSize = 8
        executor.queueCapacity = 200
        executor.setThreadNamePrefix("email-sync-")
        executor.initialize()
        return executor
    }

    @Bean(name = ["defaultTaskExecutor"])
    @Primary
    fun taskExecutor(): ThreadPoolTaskExecutor {
        val executor = ThreadPoolTaskExecutor()
        executor.corePoolSize = 4
        executor.maxPoolSize = 16
        executor.queueCapacity = 500
        executor.setThreadNamePrefix("async-")
        executor.setWaitForTasksToCompleteOnShutdown(true)
        executor.setAwaitTerminationSeconds(30)
        executor.initialize()
        return executor
    }

    @Bean(name = ["applicationTaskExecutor"])
    fun applicationTaskExecutor(
        @Qualifier("defaultTaskExecutor") executor: ThreadPoolTaskExecutor
    ): Executor = executor.threadPoolExecutor

    @Bean(name = ["taskScheduler"])
    fun taskScheduler(): TaskScheduler {
        val scheduler = ThreadPoolTaskScheduler()
        scheduler.poolSize = 4
        scheduler.setThreadNamePrefix("scheduled-")
        scheduler.setWaitForTasksToCompleteOnShutdown(true)
        scheduler.setAwaitTerminationSeconds(30)
        scheduler.initialize()
        return scheduler
    }
}
